File: C++/DesignPatterns/FileSystem_Node.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>

struct FileSystemNode {
    std::string name;
    int size = 0;
    FileSystemNode* parent = nullptr;

    explicit FileSystemNode(std::string name) : name(std::move(name)) {}

    virtual void properties() = 0;
    virtual int getSize() = 0;
    virtual void rename(const std::string& newName) = 0;
    virtual void open() = 0;
    virtual ~FileSystemNode() = default;
};

class File : public FileSystemNode {
public:
    File(const std::string& name, const std::string& content = "")
        : FileSystemNode(name), content(content) {
        size = static_cast<int>(content.size());
    }

    void properties() override {
        std::cout << "File: " << name << ", size: " << size << std::endl;
    }

    int getSize() override { return size; }

    void rename(const std::string& newName) override { name = newName; }

    void open() override { std::cout << "Opening file " << name << std::endl; }

private:
    std::string content;
};

class Folder : public FileSystemNode {
public:
    explicit Folder(const std::string& name) : FileSystemNode(name) {}

    void addChild(std::unique_ptr<FileSystemNode> node) {
        node->parent = this;
        children.push_back(std::move(node));
    }

    void properties() override {
        std::cout << "Folder: " << name << ", size: " << getSize() << std::endl;
    }

    int getSize() override {
        int total = 0;
        for (const auto& child : children) {
            total += child->getSize();
        }
        return total;
    }

    void rename(const std::string& newName) override { name = newName; }

    void open() override { std::cout << "Opening folder " << name << std::endl; }

private:
    std::vector<std::unique_ptr<FileSystemNode>> children;
};

int main() {
    auto root = std::make_unique<Folder>("root");
    root->addChild(std::make_unique<File>("readme.txt", "hello"));
    root->addChild(std::make_unique<Folder>("subfolder"));
    return 0;
}

File: Python/DesignPatterns/FileSystem_Node.py
from abc import ABC, abstractmethod
from typing import List, Optional


class FileSystemNode(ABC):
    def __init__(self, name: str) -> None:
        self.name = name
        self.size: int = 0
        self.parent: Optional["FileSystemNode"] = None

    @abstractmethod
    def properties(self) -> None:
        raise NotImplementedError

    @abstractmethod
    def get_size(self) -> int:
        raise NotImplementedError

    @abstractmethod
    def rename(self, new_name: str) -> None:
        raise NotImplementedError

    @abstractmethod
    def open(self) -> None:
        raise NotImplementedError


class File(FileSystemNode):
    def __init__(self, name: str, content: str = "") -> None:
        super().__init__(name)
        self.content = content
        self.size = len(content)

    def properties(self) -> None:
        print(f"File: {self.name}, size: {self.size}")

    def get_size(self) -> int:
        return self.size

    def rename(self, new_name: str) -> None:
        self.name = new_name

    def open(self) -> None:
        print(f"Opening file {self.name}")


class Folder(FileSystemNode):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self.children: List[FileSystemNode] = []

    def add_child(self, node: FileSystemNode) -> None:
        node.parent = self
        self.children.append(node)

    def properties(self) -> None:
        print(f"Folder: {self.name}, size: {self.get_size()}")

    def get_size(self) -> int:
        return sum(child.get_size() for child in self.children)

    def rename(self, new_name: str) -> None:
        self.name = new_name

    def open(self) -> None:
        print(f"Opening folder {self.name}")


def main() -> None:
    root = Folder("root")
    root.add_child(File("readme.txt", "hello"))
    root.add_child(Folder("subfolder"))


if __name__ == "__main__":
    main()
